### PR TITLE
handle data fetch and transformation errors and show relevant error page

### LINKED ISSUE #6891: Gracefully handle 202 / 404 response from AReS for On Demand Radio pages
**Describe the bug**
Simorgh is returning a HTTP 500 (Internal Server Error) error if we try to access an On Demand Radio page that either does not exist, or has not yet been cached by AReS. This is incorrect behaviour - Simorgh should return a HTTP 404 (Not Found) error instead

**To Reproduce**
Steps to reproduce the behaviour:

1. Go to https://www.test.bbc.com/indonesia/bbc_indonesian_radio/abcdef (non existent On Demand Radio episode)
2. Internal Server Error page displayed

In a local environment: 
1. Go to http://localhost:7080/indonesia/bbc_indonesian_radio/abcdef
2. The following message is displayed: `Cannot read property 'split' of undefined`

**Expected behaviour**
404 Not Found error page should be displayed instead

**Screenshots**
image

**Testing notes**

- [ ] This bug fix is expected to need manual testing.

**Additional context**

### PR DESCRIPTION
**Pull request by jroebu14**
Resolves #6891 

**Overall change:**
Catches Ares fetch errors/404s/202s in `fetchPageData`, logs them and then throws a Simorgh specific error that will be caught and handled further up in each route's `getInitialData` function.

The `getInitialData` function in every route now has a try/catch around the async `fetchPageData` and data transformation code to handle any data fetch and transformation errors. These error handlers return an error message `error` and error code `status` that the `withError` page HOC handles to show the correct error page to the user.

Unit tests in `src/app/routes/getInitialData.test.jsx` ensure any route added is required to implement an error handler.

**Code changes:**

- _A bullet point list of key code changes that have been made._
- _When describing code changes, try to communicate **how** and **why** you implemented something a specific way, not just **what** has changed._

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] I have added labels to this PR for the relevant pod(s) affected by these changes
- [x] I have assigned this PR to the Simorgh project

**Testing:**

- [ ] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`)
- [ ] This PR requires manual testing

### CODE REVIEW

**Comment by ryanmccombe** [comment]
You mentioned adding this to all pages, so I was going to suggest adding it here: https://github.com/bbc/simorgh/blob/43049beee73a8a095c19cb3f3ad74d59ad9a146d/src/server/index.jsx#L296

But there already is a try-catch there

I'm not 100% sure Simorgh isn't behaving as expected here - it's currently returning a 500, and will still return a 500 after this PR

File: src/app/routes/utils/fetchPageData/index.js [hunk 1]
```diff
@@ -32,22 +32,39 @@ export const getUrl = pathname => {
   return `${baseUrl}${basePath.replace(ampRegex, '')}.json${params}`; // Remove .amp at the end of pathnames for AMP pages.
 };
 
+const validateData = data => {
+  if (data) {
+    return 'content' in data;
```
Comment (karinathomasbbc) [inline src/app/routes/utils/fetchPageData/index.js]: I've never seen this syntax before - what does it do (if you don't mind me asking?)

File: src/app/lib/logger.const.js [hunk 2]
```diff
@@ -7,6 +7,7 @@ const logCodes = {
   DATA_FETCH_ERROR: 'data_fetch_error',
   DATA_NOT_FOUND: 'data_response_404',
   DATA_REQUEST_RECEIVED: 'data_request_received',
+  DATA_PROCESSING_ERROR: 'data_processing_error',
```
Comment (karinathomasbbc) [inline src/app/lib/logger.const.js]: I couldn't see where this was being used - we might not need it...

File: src/app/routes/getInitialData.test.jsx [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        JSON.stringify({
+          dataIsNotAsExpected: 'dataIsNotAsExpected',
+        }),
+      );
+
+      const actual = await getInitialData(MOCK_PATH);
+
+      expect(actual.status).toEqual(502);
+      expect(actual.error.toString()).toMatch(
+        'Error: Unexpected data format in response when requesting',
```
Comment (karinathomasbbc) [inline src/app/routes/getInitialData.test.jsx]: Could / should a URL be included in the error message?

File: src/app/routes/utils/fetchPageData/index.js (hunk 1 above)
Comment (jroebu14) [reply src/app/routes/utils/fetchPageData/index.js]: I do mind you asking it's a very private matter to me 😆 joking. So it checks for the `content` key in the object `data`. True if it has one and false if not

Comment (karinathomasbbc) [reply src/app/routes/utils/fetchPageData/index.js]: thank you for explaining!

**Comment by ryanmccombe** [comment]
Some thoughts on a radically different approach that might solve these problems, and more problems we're likely to run into in the future as we add more `getInitialData` functions, and more complexity to them

I wonder if `getInitialData` should be just part of the platform, rather than having each page make their own?  It's hard to generalise error handling (or much at all) when every page implements this function from scratch

Having a generalised `getInitialData` just part of the standard flow of all pages would eliminate quite a few problems, I think

Some pages have slightly different requirements, but we could expose hooks / events to allow different pages the ability to implement those differences.

Eg, Radio pages pluck individual fields from the `pageData` rather than passing down the entire JSON to the page.

But this doesn’t mean that radio pages need to re-implement all data fetching, we can just allow the platform to accept a redux-esque `mapPageDataToProps` function to handle that, as well as any other options that change per page.

This minimises the duplicate code in every page to be just what is actually unique for that page, and also provides us with a standard place to implement `getInitialData` behaviour that should be, by default, shared across all pages, such as error handling

eg, the routing API could be something like:

```javascript
// Removal of getInitialData - just use the standard function
{
  component: ArticlePage,
  path: articlePath,
  exact: true
}
```

```javascript
// Pass in options for how getInitialData should happen
{
  component: RadioPage,
  path: radioPath,
  exact: true,
  getInitialData: {
    url: addRendererEnvOnTest(),
    mapPageDataToProps: pageData => ({ id: pageData.id })
  }
}
```

**Comment by rebeccamcginn** [comment]
Perhaps we could have a tech design meeting to discuss this? I'd be keen to hear the discussion

**Comment by jroebu14** [comment]
I've always said we shouldn't over provide our components with unprocessed and complex data so I'm not too into the idea of a function that by default passes all data to a component. I know article pages do this but I don't think it should. Also limiting fetching data to one url means we can't fetch from multiple sources.

**Comment by ryanmccombe** [comment]
> I've always said we shouldn't over provide our components with unprocessed and complex data so I'm not too into the idea of a function that by default passes all data to a component. I know article pages do this but I don't think it should.

It's not just articles - 4-5 `getInitialData` functions do this.  This approach would reduce the instances of this pattern from five to one - and make it easier to get to zero 

> Also limiting fetching data to one url means we can't fetch from multiple sources.

Our current approach of copying and pasting `getInitialData` functions doesn't allow us to fetch from multiple sources for free either.  If we ever needed to do that, we'd have to implement it.

If we had a shared place to implement it, that option would become available to all pages, not just one.

```javascript
{
  component: RadioPage,
  path: radioPath,
  exact: true,
  getInitialData: {
    urls: ['url-1', 'url-2']
  }
}
```

**Review by ryanmccombe — APPROVED** [review]
👍

**Review by karinathomasbbc — COMMENTED** [review]
👍

**Review by karinathomasbbc — APPROVED** [review]

**Pull request merged**